src/python/pants/engine/legacy/graph.py

```diff
@@ -511,19 +519,44 @@ def transitive_fingerprinted_targets(
     for dep in tht.dependencies
   )
 
-  fingerprinted_targets = tuple(
-    TransitiveFingerprintedTarget(
-      was_root=was_root,
-      address=tht.root.address,
-      intransitive_fingerprint_arg=tht.root.adaptor.intransitive_fingerprint,
-      transitive_fingerprint_arg=stable_json_sha1(tuple(
-        dep.root.adaptor.intransitive_fingerprint for dep in tht.dependencies
-      )),
-    )
-    for tht, was_root in transitive_hydrated_targets
-  )
+  intransitive_fingerprint_dict = {}
+  visited_targets = set()
+  fingerprinted_targets = []
+  for tht, was_root in transitive_hydrated_targets:
+    if tht.root.address in visited_targets:
+      continue
 
-  yield FingerprintedTargetCollection(fingerprinted_targets)
+    intransitive_fingerprint = intransitive_fingerprint_dict.get(tht.root.address, None)
+    if not intransitive_fingerprint:
+      sources = getattr(tht.root.adaptor, 'sources', None)
+      sources_snapshot = sources.snapshot if sources else None
+      intransitive_fingerprint = TransitiveFingerprintedTarget.calculate_intransitive_fingerprint_for_target_adaptor(
+        tht.root.adaptor, sources_snapshot)
+      intransitive_fingerprint_dict[tht.root.address] = intransitive_fingerprint
+
+
+    for dep in tht.dependencies:
+      dep_intransitive_fingerprint = intransitive_fingerprint_dict.get(dep.root.address, None)
+      if not dep_intransitive_fingerprint:
+        dep_sources = getattr(dep.root.adaptor, 'sources', None)
```

Comment:
Nit: This block and the one before looks like it could be extracted into a common function

Reply:
Done now via `@memoized_classproperty!